tracer/src/Datadog.Trace/Iast/ReturnedHeadersAnalyzer.cs

```diff
@@ -39,6 +41,140 @@ internal static void Analyze(IHeaderDictionary responseHeaders, IntegrationId in
         AnalyzeXContentTypeOptionsVulnerability(responseHeaders, integrationId, serviceName, responseCode);
         AnalyzeStrictTransportSecurity(responseHeaders, integrationId, serviceName, responseCode, protocol);
         AnalyzeUnvalidatedRedirect(responseHeaders, integrationId);
+        AnalyzeHeaderInjectionVulnerability(responseHeaders, integrationId);
+    }
+
+    // In header injections, we should exclude some headers to prevent false positives:
+    // location: it is already reported in UNVALIDATED_REDIRECT vulnerability detection.
+    // Sec-WebSocket-Location, Sec-WebSocket-Accept, Upgrade, Connection: Usually the framework gets info from request
+    // access-control-allow-*: when the source of the tainted range is the request header origin or access-control-request-headers
+    // set-cookie: We should ignore set-cookie header if the source of all the tainted ranges are cookies
+    // We should exclude the injection when the tainted string only has one range which comes from a request header with the same name that the header that we are checking in the response.
+    // Headers could store sensitive information, we should redact whole <header_value> if:
+    // <header_name> matches with a RegExp
+    // <header_value> matches with  a RegExp
+    // We should redact the sensitive information from the evidence when:
+    // Tainted range is considered sensitive value
+
+#if NETFRAMEWORK
+    private static void AnalyzeHeaderInjectionVulnerability(NameValueCollection responseHeaders, IntegrationId integrationId)
+#else
+    private static void AnalyzeHeaderInjectionVulnerability(IHeaderDictionary responseHeaders, IntegrationId integrationId)
+#endif
+    {
+        try
+        {
+            IastModule.OnExecutedSinkTelemetry(IastInstrumentedSinks.HeaderInjection);
+
+            var tracer = Tracer.Instance;
+            var currentSpan = (tracer.ActiveScope as Scope)?.Span;
+            var traceContext = currentSpan?.Context?.TraceContext;
+
+#if NETFRAMEWORK
+            foreach (var headerKey in responseHeaders.AllKeys)
+#else
+            foreach (var headerKey in responseHeaders.Keys)
+#endif
+            {
+                if (string.IsNullOrWhiteSpace(headerKey))
+                {
+                    continue;
+                }
+
+                bool isHeaderInjectionException = false;
+                foreach (var excludeType in headerInjectionExceptions)
+                {
+                    if (excludeType.Equals((string)headerKey, StringComparison.OrdinalIgnoreCase))
+                    {
+                        isHeaderInjectionException = true;
+                        break;
+                    }
+                }
+
+                if (isHeaderInjectionException)
+                {
+                    continue;
+                }
+
+                string headerValue = responseHeaders[headerKey];
```

Comment:
In .NET Core, the return value is `StringValues`. By casting to a `string`, you will throw an exception if there are actually _multiple_ header values. 

A better option is to handle the possibility of there being more than one value set, and handle it appropriately. Luckily, I wrote a blog post about this 3 days ago 😂 https://andrewlock.net/a-brief-look-at-stringvalues/

Also, just FYI, if the same header is set multiple times in .NET FX `NameValueCollection` will return them here using `string.Join(",", values)`. That _may_ be ok for you here, but I don't know if you have anything in the spec about how to handle this situation (i.e. the header is set more than once)?

Reply:
I have updated the code to handle multiple values in one header for both netcore and net framework. In Net framework, I have used GetValues to get all the values and avoid the join. I have also added integration tests that cover this case. Thank you!

Reply:
> In Net framework, I have used GetValues to get all the values and avoid the join.

So there's definitely a question of whether we _should_ do that tbh... By using `GetValues()` we're _always_ allocating an array for every accessed header. I wonder if it's actually a problem or not if we use `GetValue()` and join the values with a `','`? Happy with whatever you decide, just wanted to flag it 🙂

Reply:
The problem with the "," separated values is that we loose the tainting since the join is done in libraries that we don't instrument, so we would not be able to know if the value is tainted.